Fix sun rendering in splitscreen

Berthalamew opened the pull request:
- Fixes bug with screen effects breaking when gamma is disabled from the following PR: #638
![image](https://github.com/pnill/cartographer/assets/40469582/cf670b1b-1b96-4fae-a843-8efb9486d1e3)

Berthalamew committed:
add lens flare definition

Berthalamew committed:
misc changes

- add color conversion function and pixel32 globals
- add offset for datum_try_and_get
- add functions to get global window bounds

Berthalamew committed:
rewrite fullscreen effect functions and sun drawing

- sun drawing is unfinished for splitscreen
- c_rasterizer_constant_4f_cache is unfinished

Berthalamew committed:
add NONE checks to first person interpolation code

Berthalamew committed:
update render_projection_point_to_screen logic

move calculations so they aren't done if v2 is less than 0.f

Berthalamew committed:
update scenario structs

- add scenario_interpolator_get_result function

Berthalamew committed:
add screen effect structures

Berthalamew committed:
add rasterizer perf event functions

Berthalamew committed:
add perf events

- update parameter name in rasterizer_dx9_draw_rect

Berthalamew committed:
misc updates to rasterizer dx9 code

- update enum
- add perf event to water refraction

Berthalamew committed:
sun fix and screen effects rewrite

Berthalamew committed:
add some real math functions

Berthalamew committed:
add depth of field functions

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9.h:
```diff
@@ -0,0 +1,70 @@
+            int32 i = 0;
+            for (;
+                !(
+                    (uint32)values[index].i ^ (uint32)constants[i].i |
```
you convert this value from float to uint, losing the value itself. you need to dereference it as uint, to preserve the actual value...

Berthalamew replied:
fixed

Berthalamew committed:
fix float comparisons in c_rasterizer_constant_4f_cache

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_screen_effect.cpp:
```diff
@@ -1,8 +1,217 @@
+	rasterizer_dx9_apply_gamma_and_brightness(target);
+	rasterizer_dx9_perf_event_end();
+
+	global_d3d_device->StretchRect(global_d3d_surface_render_primary_get(), NULL, global_d3d_surface_render_resolved_get(), NULL, D3DTEXF_NONE);
```
I don't understand the extra copy here

Berthalamew replied:
The copy was originally done at the start of the function.
I moved it to the bottom once the primary render surface had the sun effect written to it.

nukeulater replied:
but why are you copying it to the resolved surface if the resolved surface gets copied in the back-buffer during `rasterizer_dx9_apply_gamma_and_brightness()`?

Berthalamew replied:
It copies it from target_25 to the back-buffer in rasterizer_dx9_apply_gamma_and_brightness?

nukeulater replied:
it copies it from target 37, which is the "resolved" surface

Berthalamew replied:
I debugged it and it's copying it from 25.

nukeulater replied:
yeah, 25 hex..., which is 37 in decimal

Berthalamew committed:
add changes and work from nuke

Berthalamew committed:
render changes and additions

- add perf events for sun draw
- add dx9 rasterizer functions

nukeulater commented on xlive/Blam/Engine/render/render_cameras.cpp:
```diff
@@ -43,3 +45,63 @@ void __cdecl render_camera_build_viewport_frustum_bounds(const s_camera* camera,
+				+ obb->matrix[3][0];			
+
+
+			rect->x0 = (v6 / divisor);
```
this is a vector4d, also a better name would be out_view_position

Berthalamew committed:
add rasterizer globals structure

- remove unneeded functions
- adjust naming in render_projection_point_to_screen

nukeulater commented on xlive/Blam/Engine/rasterizer/rasterizer_globals.h:
```diff
@@ -1,5 +1,119 @@
+	bool use_d3d9_ex;
+	int8 pad[2];
+	decltype(Direct3DCreate9Ex)* d3d9_create_ex_proc;
+	uint32 resolution_x;
```
point2d?

Berthalamew replied:
point2d uses int16's while these are 32 bit ints

Berthalamew committed:
misc changes

- add more assert macros
- add machine flags
- move real_vector4d union

Berthalamew committed:
Fix sun in splitscreen

- rewrite render target initialization code
- add string parameter to rasterizer_dx9_perf_event_end
- rewrite convolution shader and use new convolution function

Berthalamew committed:
remove unneeded fixme comment

Berthalamew committed:
misc changes

Berthalamew committed:
rewrite rasterizer_suns_glow_draw function

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.cpp:
```diff
@@ -72,4 +190,775 @@ void __cdecl rasterizer_set_render_target_internal_hook_set_viewport(IDirect3DSu
+// We really shouldn't do this but we have to in order to "own" the primary render targets...
+void rasterizer_dx9_patch_globals(void)
+{
+    WriteValue(Memory::GetAddress(0x25EE53 + 1), &global_d3d_surface_sun_glow_primary);
```
no.

Berthalamew replied:
removed

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.cpp:
```diff
@@ -72,4 +190,775 @@ void __cdecl rasterizer_set_render_target_internal_hook_set_viewport(IDirect3DSu
+
+void rasterizer_dx9_get_render_target_surface_patch_calls(void)
+{
+    PatchCall(Memory::GetAddress(0x19A90F), rasterizer_dx9_get_render_target_surface);
```
either global detours or nothing (unless you need a hook like this in a specific location, which is alright). stop doing these hooks this way, clutters the code.

nukeulater replied:
also fucks with the references when debugging with IDA

Berthalamew replied:
removed and using WriteJmpTo now

nukeulater replied:
why not DETOUR_ATTACH?

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_dof.cpp:
```diff
@@ -0,0 +1,130 @@
+    */
+}
+
+void rasterizer_dx9_draw_convolution_rect(uint32 screen_resolution_x, uint32 screen_resolution_y, D3DCOLOR color)
```
this just draws a single color quad, has nothing to do with convolution or anything in particular with convolution

Berthalamew replied:
rasterizer_dx9_draw_overlay_rect?

Berthalamew committed:
rename functions

- rasterizer_dx9_render_screen_effects to rasterizer_dx9_postprocess_scene
- rasterizer_dx9_draw_convolution_rect to rasterizer_dx9_draw_overlay_rect

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_lens_flares.cpp:
```diff
@@ -0,0 +1,443 @@
+    real_vector4d sun_center;
+    if (render_projection_point_to_screen(&position, definition->occlusion_radius, &sun_center, &bounds))
+    {
+        const real32 viewport_width = rectangle2d_width(&global_window_parameters->camera.viewport_bounds);
```
int16 in original code

Berthalamew replied:
changed

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_lens_flares.cpp:
```diff
@@ -0,0 +1,443 @@
+
+            rasterizer_dx9_set_vertex_shader_permutation(10);
+
+            real32 viewport_middle_x = ((real32)viewport_width / 2.f);
```
this division is done with integers

Berthalamew replied:
changed

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.cpp:
```diff
@@ -1,22 +1,96 @@
+
+real32 g_sun_size = 0.2f;
+
+IDirect3DTexture9* g_backbuffer_texture = NULL;
```
tried to find an argument on why these should stay, and I couldn't find anything reasonable.

Berthalamew replied:
Going to remove all the ones that we're not currently using in the re-written functions.

nukeulater replied:
shouldn't be using any of these because we could already access them from the game itself.

Berthalamew replied:
There's no point going back and forth between the dll and the game if we already have the functions reversed that use them.
It's extra functions rather than having the global accessed locally.

nukeulater replied:
that's why I proposed mapping the game globals inside a structure for easy access?

Berthalamew replied:
At the end of the day it's a function call that isn't really needed when the global can be used.

Berthalamew committed:
add suggested changes

- remove global surfaces and textures that aren't used by our functions
- use int16's for math in glow calculations instead of floats

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.h:
```diff
@@ -67,18 +67,18 @@ struct s_rasterizer_target
 
 extern IDirect3DTexture9* g_backbuffer_texture;
-extern IDirect3DSurface9* g_backbuffer_surface;
+// extern IDirect3DSurface9* g_backbuffer_surface;
```
```cpp
struct s_rasterizer_data
{
	void** field_0;
	void** field_4;
	void** field_8;
	void** field_C;
	uint8 gap_10[4];
	void* bloom_globals;
	IDirect3D9* d3d;
	uint8 gap_1C[8];
	IDirect3DSurface9* d3d_surface_render_primary;
	IDirect3DSurface9* d3d_surface_render_primary_z;
	IDirect3DSurface9* d3d_surface_render_z_as_target_z;
	IDirect3DBaseTexture9* d3d_texture_render_resolved;
	IDirect3DSurface9* d3d_surface_render_resolved;
	IDirect3DSurface9* d3d_surface_screenshot;
	uint8 gap_3C[4];
	IDirect3DBaseTexture9* d3d_texture_backbuffer;
	IDirect3DSurface9* d3d_surface_backbuffer_depth;
	uint8 gap_48[8];
	IDirect3DBaseTexture9* d3d_texture_sun_glow_primary;
	IDirect3DSurface9* d3d_surface_sun_glow_primary;
	IDirect3DBaseTexture9* d3d_texture_sun_glow_secondary;
	IDirect3DSurface9* d3d_surface_sun_glow_secondary;
	IDirect3DBaseTexture9* d3d_texture_motion_sensor;
	IDirect3DSurface9* d3d_surface_motion_sensor;
	uint8 gap_68[4];
	IDirect3DSurface9* field_6C;
	IDirect3DSurface9* field_70;
	IDirect3DSurface9* field_74;
	IDirect3DSurface9* field_78;
	IDirect3DSurface9* field_7C;
	IDirect3DSurface9* field_80;
	IDirect3DSurface9* field_84;
	uint8 gap_88[4];
	void* d3d_device;
	uint8 gap_90[8];
};
ASSERT_STRUCT_SIZE(s_rasterizer_data, 152);

s_rasterizer_data* rasterizer_get_data()
{
    return Memory::GetAddress<s_rasterizer_data*>(0xA3C628);
}

rasterizer_get_data()->d3d_surface_render_primary;

```

you can also do:

```cpp
&rasterizer_get_data()->d3d_surface_render_primary;
```
and initialize the targets this way
```cpp
global_d3d_device->CreateRenderTarget(screen_bounds_width, screen_bounds_height, format, type, quality, false, &rasterizer_get_data()->d3d_surface_render_z_as_target_z, NULL);
```
and avoid all of these double pointer calls: 

```cpp
IDirect3DSurface9** global_d3d_surface_render_primary_get(void);
```

nukeulater replied:
basically a structure mapping game's memory

Berthalamew replied:
Not a fan of doing this _at this moment_.
Could create confusion when referencing later as it'd be broken up into this struct and our globals, as well as any future render targets or additions we make.

nukeulater replied:
Personally not confused at all. I'm a big fan of doing this, exactly at this moment, compared to what is currently in.

nukeulater commented on xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.cpp:
```diff
@@ -1,22 +1,89 @@
+void rasterizer_dx9_targets_apply_patches(void)
 {
-	return Memory::GetAddress<e_rasterizer_target*>(0xA3E228);
+    WriteJmpTo(Memory::GetAddress(0x25FDC0), rasterizer_dx9_set_target);
```
`DETOUR_ATTACH` is preferred over `WriteJmpTo`

Berthalamew replied:
Changed

Berthalamew committed:
change WriteJmpTo to DETOUR_ATTACH

nukeulater commented on xlive/Blam/Engine/rasterizer/rasterizer_globals.h:
```diff
@@ -1,5 +1,119 @@
+ASSERT_STRUCT_SIZE(s_rasterizer_fullscreen_parameters, 16);
+
+// TODO: Finish this, some data above might be included in it
+struct s_rasterizer_globals
```
by the same logic, this should also be removed?

Berthalamew replied:
This is a struct that exists in the original game.
After taking another look, the globals above it are related to shader submit logic.

Berthalamew committed:
remove TODOs

- globals above are for shader submit globals and are not a part of s_rasterizer_globals

Berthalamew committed:
create memory map for main dx9 globals

Berthalamew committed:
re-add rasterizer_dx9_device_get_interface
